### PR TITLE
add RetinaNet network

### LINKED ISSUE #4383: Add RetinaNet network
**Is your feature request related to a problem? Please describe.**
Add RetinaNet network

**Describe the solution you'd like**
Add RetinaNet network

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Can-Zhao**
Signed-off-by: Can Zhao <canz@nvidia.com>

Fixes #4383.
add retinanet

### Description
add retinanet networks

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/apps/detection/networks/base_detection_network.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+from abc import abstractmethod
+from typing import Sequence, Union
+
+from torch import Tensor, nn
+
+from monai.utils import ensure_tuple_rep
+from monai.utils.module import look_up_option
+
+
+class BaseDetectionNetwork(nn.Module):
```
Comment (wyli) [inline monai/apps/detection/networks/base_detection_network.py:21]: Hi @Can-Zhao this kind of `BaseDetectionNetwork` generic abstraction will need more discussions, could you please remove it for now,

**Review by wyli — APPROVED** [review]
New net arch with 2/3d support, torchscript support not tested.

File: tests/test_retinanet.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+]
+
+TEST_CASES = []
+for case in [TEST_CASE_1, TEST_CASE_2, TEST_CASE_3, TEST_CASE_2_A, TEST_CASE_3_A]:
+    for model in [resnet10, resnet18, resnet34, resnet50, resnet101, resnet152, resnet200]:
+        TEST_CASES.append([model, *case])
+
+
+@unittest.skipUnless(has_torchvision, "Requires torchvision")
+class TestRetinaNet(unittest.TestCase):
```
Comment (Nic-Ma) [inline tests/test_retinanet.py:94]: As we are going to develop a detection bundle soon, could you please help add a test for TorchScript export?
I will disable the auto-merge for a while.

Thanks in advance.

File: monai/apps/detection/networks/retinanet_network.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+    A classification head for use in RetinaNet.
+
+    This head takes a list of feature maps as inputs, and outputs a list of classification maps.
+    Each output map has same spatial size with the corresponding input feature map,
+    and the number of output channel is num_anchors * num_classes.
+
+    Args:
+        in_channels: number of channels of the input feature
+        num_anchors: number of anchors to be predicted
+        num_classes: number of classes to be predicted
```
Comment (Nic-Ma) [inline monai/apps/detection/networks/retinanet_network.py:40]: missing doc-string for `spatial_dims` and `prior_probability`(also missing typehint).

Thanks.

File: monai/apps/detection/networks/retinanet_network.py [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+        backbone: a ResNet model, used as backbone.
+        spatial_dims: number of spatial dimensions of the images. We support both 2D and 3D images.
+        pretrained_backbone: whether the backbone has been pre-trained.
+        trainable_backbone_layers: number of trainable (not frozen) resnet layers starting from final block.
+            Valid values are between 0 and 5, with 5 meaning all backbone layers are trainable.
+            When pretrained_backbone is False, this value is set to be 5.
+            When pretrained_backbone is True, if ``None`` is passed (the default) this value is set to 3.
+        returned_layers: returned layers to extract feature maps. Each returned layer should be in the range [1,4].
+            len(returned_layers)+1 will be the number of extracted feature maps.
+            There is an extra maxpooling layer LastLevelMaxPool() appended.
```
Comment (Nic-Ma) [inline monai/apps/detection/networks/retinanet_network.py:329]: missing doc-string for `trainable_backbone_layers`.

Thanks.

**Review by Nic-Ma — APPROVED** [review]
It overall looks good to me.
Put some minor comments inline.
I will disable the auto-merge for a while.

Thanks.

**Comment by Can-Zhao** [comment]
> New net arch with 2/3d support, torchscript support not tested.

torchscript support has been tested now

File: tests/test_retinanet.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        backbone = model(**input_param).to(device)
+        test_script_save(backbone, data)
+        feature_extractor = resnet_fpn_feature_extractor(
+            backbone=backbone,
+            spatial_dims=input_param["spatial_dims"],
+            pretrained_backbone=input_param["pretrained"],
+            trainable_backbone_layers=None,
+            returned_layers=[1, 2],
+        ).to(device)
+        # test_script_save(feature_extractor, data)
```
Comment (Nic-Ma) [inline tests/test_retinanet.py:153]: Please remove this commented code.

Thanks.

**Pull request merged**